### ISSUE TITLE
With vfio_mode=vfio and all VFIO devices will be in the same VFIO group in the guest

### ISSUE THREAD

**Issue by dgibson** [kata-containers/kata-containers#2899]
# Description of problem

[This is a theoretical bug, I haven't actualy tested the behaviour yet]

With the default `hotplug_vfio_on_root_bus=false`, then all VFIO devices in a container will be hotplugged behind the same PCIe to PCI bridge.  This means that they will all appear in the same VFIO group within the guest.

This is ok for what I suspect is the most common use case: a single DPDK application which consumes all the VFIO devices and attaches them to the same container. However it will fail for any application which attempts to attach different devices to different containers (since a VFIO group / IOMMU group is the unit of ownership for VFIO containers).

# Expected result

VFIO devices which are in different groups in the host should be in different groups in the guest.

# Actual result

All VFIO devices will be in the same group in the guest, regardless of host topology.

**Comment by eadamsintel** [comment]
I am having an issue that a virtual function on the host is in its own IOMMU group but when hot plugged into the QEMU VM it shares an IOMMU group with the PCI-PCI bridge. If it is cold booted in the VM then it ends up in its own IOMMU group.  Would this patch solve that issue or would a hotplugged VF share an IOMMU group with the PCI-PCI bridge?  TIA!

**Comment by dgibson** [comment]
@eadamsintel yes, that has the same root cause as this bug.  Classic PCI (not PCIe) doesn't have requestor IDs, so the IOMMU can't distinguish between devices behind a PCIe to PCI bridge.  They all appear with the bridge's requestor ID, so they all show up in the same group.

Solving this bug will require a different arrangement for plugging the devices - probably each device on it's own PCIe root port (except for host multi device groups, which would need to go behind a PCIe to PCI bridge itself hotplugged into a PCIe root port).

That said, sharing an IOMMU group with *just* the bridge should generally be ok.  As long as the group isn't shared with any non-bridge devices you should be able to use VFIO ok within the guest.

**Comment by eadamsintel** [comment]
@dgibson I have an actual use case right now that I think this would fix. My vf gets hot plugged into the same IOMMU group as a PCI-PCI bridge but, due to security reasons, the vf driver does not allow any other devices in that IOMMU group. I think your work on vfio would solve this at least based on your description.  I thought this was a hypervisor limitation and not something the runtime or agent could fix.  I do look forward to trying out your solution when it is ready.

**Comment by dgibson** [comment]
> @dgibson I have an actual use case right now that I think this would fix. My vf gets hot plugged into the same IOMMU group as a PCI-PCI bridge but, due to security reasons, the vf driver does not allow any other devices in that IOMMU group. I think your work on vfio would solve this at least based on your description. I thought this was a hypervisor limitation and not something the runtime or agent could fix. I do look forward to trying out your solution when it is ready.

First, note that this issue is specifically about the new `vfio_mode=vfio` behaviour, where the device is presented as a VFIO device within the container.  Since you're discussing a vf driver, I'm guessing you're using the old `vfio_mode=guest-kernel` 
behaviour, where a custom driver in the guest kernel is attached to the device.  Changes that fix this behviour for `vfio_mode=vfio` *might* also change it for `vfio_mode=guest-kernel`, but they might not, depends what they end up looking like.

Second, I'm unlikely to be tackling this problem any time soon - it's not particularly relevant for the use cases I'm most interested in.

The good news is, you can probably work around this right now: if you set `hotplug_vfio_on_root_bus = true` and `pcie_root_port = 1` (or a larger number if you have multiple VFs) in your configuration, it should put the VFs onto root ports rather than under a PCIe to PCI bridge, which means they should be in separate IOMMU groups.  There are a number of issues with the root port behaviour at present, but I believe it will suffice for your needs here.
